byzcoin/contract_deferred.go

```diff
@@ -0,0 +1,404 @@
+		c.DeferredData.ProposedTransaction.Instructions[index].SignerIdentities = append(c.DeferredData.ProposedTransaction.Instructions[index].SignerIdentities, identity)
+		c.DeferredData.ProposedTransaction.Instructions[index].Signatures = append(c.DeferredData.ProposedTransaction.Instructions[index].Signatures, signature)
+		// Save and send the modifications
+		cosiDataBuf, err2 := protobuf.Encode(&c.DeferredData)
+		if err2 != nil {
+			return nil, nil, errors.New("couldn't encode DeferredData")
+		}
+		sc = append(sc, NewStateChange(Update, inst.InstanceID,
+			ContractDeferredID, cosiDataBuf, darcID))
+		return
+	case "execProposedTx":
+		// This invocation tries to execute the transaction stored with the
+		// "Spawn" invocation. If it is successful, this invocation fills the
+		// "ExecResult" field of the "deferredData" struct.
+
+		// In the following we are creating a new StagingStateTrie from the
+		// readonly state try by copying the data.
+		nonce, err2 := rst.GetNonce()
+		if err2 != nil {
+			return nil, nil, errors.New("couldn't get the nonce: " + err2.Error())
+		}
+		sst, err2 := newMemStagingStateTrie(nonce)
+		if err2 != nil {
+			return nil, nil, errors.New("Failed to created stagingStateTrie: " + err2.Error())
+		}
+		rst.ForEach(sst.Set)
+
+		instructionIDs := make([][]byte, len(c.DeferredData.ProposedTransaction.Instructions))
+
+		for i, proposedInstr := range c.DeferredData.ProposedTransaction.Instructions {
+
+			// In case it goes well, we want to return the proposed Tx InstanceID
+			instructionIDs[i] = proposedInstr.DeriveID("").Slice()
+
+			instructionType := proposedInstr.GetType()
+
+			var contractID string
+			switch instructionType {
+			case SpawnType:
+				contractID = proposedInstr.Spawn.ContractID
+			case InvokeType:
+				contractID = proposedInstr.Invoke.ContractID
+			case DeleteType:
+				contractID = proposedInstr.Delete.ContractID
+			}
+
+			fn, exists := c.s.GetContractConstructor(contractID)
+			if !exists {
+				return nil, nil, errors.New("couldn't get the root function")
+			}
+			rootInstructionBuff, err := protobuf.Encode(&proposedInstr)
+			if err != nil {
+				return nil, nil, errors.New("couldn't encode the root instruction buffer")
+			}
+			contract, err := fn(rootInstructionBuff)
+			if err != nil {
+				return nil, nil, errors.New("couldn't get the root contract")
+			}
+			err = contract.VerifyDeferredInstruction(sst, proposedInstr, c.DeferredData.InstructionHashes[i])
+			if err != nil {
+				return nil, nil, fmt.Errorf("verifying the instruction failed: %s", err)
+			}
+
+			var stateChanges []StateChange
+			switch instructionType {
+			case SpawnType:
+				stateChanges, _, err = contract.Spawn(sst, proposedInstr, coins)
+			case InvokeType:
+				stateChanges, _, err = contract.Invoke(sst, proposedInstr, coins)
+			case DeleteType:
+				stateChanges, _, err = contract.Delete(sst, proposedInstr, coins)
+
+			}
+
+			sst.StoreAll(stateChanges)
```

Comment:
do you have a test that checks whether the deferred contract works with instructions that depend on each other (all in the same transaction)? i glanced the tests but didn't see one. maybe use the value contract to set a value and then in a different instruction try to get the same value.

Reply:
Added one. The simplest case I could implement is deleting a value contract then trying to call it and see if I get an error.